new CI tutorial -- publish allure report -- initial commit DOC-2491 (#498)
harness/developer-hub: tutorials/build-code/ci-publish-allure-report.md

---
sidebar_position: 120

> description: Create a CI pipeline that runs a maven test and generates an allure report that you can view in the Harness UI.

Reviewer: ```suggestion
description: Create a CI pipeline that runs a Maven test and generates an Allure Report that you can view in the Harness UI.
```

keywords: [Hosted Build, Continuous Integration, Hosted, CI Tutorial, maven, Allure]
---

> # Publish an allure report to the Artifacts tab 

Reviewer: ```suggestion
# Publish an Allure Report to the Artifacts tab 
```

> This tutorial provides an example pipeline that builds a Java maven application and generates an allure report that you can view in the Harness UI. 

Reviewer: ```suggestion
This tutorial provides an example pipeline that builds a Java Maven application and generates an Allure Report that you can view in the Harness UI. 
```

:::note

> Artifact tab publishing is supported in Kubernetes infrastructure only.

Reviewer: ```suggestion
Artifact tab publishing is supported in the Kubernetes infrastructure only.
```

:::

### Pipeline workflow 

This pipeline has five steps that do the following: 

1. Runs the Maven tests present in the project. 

> 2. Generates the allure report using the `allure-results` generated as part of step 1. The Allure tool is present in the `solutis/allure:2.9.0` docker image used as part of the step.

Reviewer: ```suggestion
2. Generates the Allure Report using the `allure-results` generated as part of step 1. The Allure tool is present in the `solutis/allure:2.9.0` Docker image used as part of the step.
```

> 3. Combines the allure report to a single HTML file.

Reviewer: ```suggestion
3. Combines the Allure Report into a single HTML file.
```

>    Viewing an allure report in browser requires running a web server via `allure open` command. But this won't allow viewing the report after the CI pipeline is complete.

Reviewer: ```suggestion
   Viewing an Allure Report in a browser requires running a web server via the `allure open` command. But this won't allow viewing the report after the CI pipeline is complete.
```

>    We will use [allure-combine](https://pypi.org/project/allure-combine/) tool to convert the allure report into a single HTML file. Running `allure-combine .` inside `allure-report` will generate `complete.html` file. This step uses `shubham149/allure-combine:latest` docker image with `allure-combine` tool present in it.

Reviewer: ```suggestion
   Use the [allure-combine](https://pypi.org/project/allure-combine/) tool to convert the Allure Report into a single HTML file. Running `allure-combine .` inside `allure-report` generates the `complete.html` file. This step uses the  `shubham149/allure-combine:latest` Docker image with the `allure-combine` tool present in it.
```

4. Publishes the report to a  Google Cloud Storage (GCS) bucket.
 

> 5. Uses static web hosting in GCS to display the allure report in the **Artifacts** tab. 

Reviewer: ```suggestion
5. Uses static web hosting in GCS to display the Allure Report on the **Artifacts** tab. 
```

>    `plugins/artifact-metadata-publisher` plugin adds the input urls to artifact tab in Harness UI.

Reviewer: ```suggestion
   The `plugins/artifact-metadata-publisher` plugin adds the input URLs to the **Artifacts** tab in the Harness UI.
```

1. Create a file secret with your GCP authorization credentials. Do the following:

  a. Run this command on the VM:
    ```
    gcloud auth application-default login
    ```
    This creates the following credentials file:  
    `/home/$(whoami)/.config/gcloud/application_default_credentials.json`
  
  b. Create a [Harness secret](/docs/platform/6_Security/3-add-file-secrets.md) for this file. 

2. Create the following connectors if you don't have them:

   - [GitHub Connector](/docs/platform/7_Connectors/connect-to-code-repo.md)
   - [GCP Connector](/docs/platform/7_Connectors/connect-to-google-cloud-platform-gcp.md) 
      Use the file secret you just created for the GCP credentials.
   - [Docker Hub Connector](/docs/platform/7_Connectors/ref-cloud-providers/docker-registry-connector-settings-reference.md)  
      You can choose to download Harness images from the [Harness Image Registry](docs/platform/7_Connectors/connect-to-harness-container-image-registry-using-docker-connector.md) instead of Docker Hub.

3.  Create a [publicly available bucket in GCS](https://cloud.google.com/storage/docs/access-control/making-data-public#objects) to store the report. 

> 4.  Update the report URL in steps 4 and 5 with your GCS bucket name ($GCS_BUCKET_NAME).

Reviewer: "in steps 4 and 5" - steps 4 and 5 where? In the Pipeline workflow section? It's not clear. Please revise for clarity.
